### ISSUE TITLE
Is a goal of these tests to display best performance or mediocre performance?

### ISSUE THREAD

**Issue by name99-org** [scalable-analyses/sme#2]
Is the goal of these tests to show optimal performance, or performance in the hands of non-experts?
Of course both types of tests have value. 

The reason I ask is that even I, with very limited knowledge of SME, can see many performance issues in the code you are benchmarking. These include

- use of Z rather than ZA as target register in the SSVE operations
- non-use of multi-vector (VG4x) loads and stores
- vastly sub-optimal construction of transposed matrices or column loads (should multi-vector strided load to Z, move to ZA horizontals, move back to Z as ZA verticals)

If you are benchmarking for the low-end, I think it's probably a good idea to indicate in your code best practice, so as to slowly lift general coding standards.
There has been some difficulty knowing exactly what's feasible or not given the lack of official SME2 guidelines, but this changes with 
https://developer.arm.com/documentation/109246/0100/Introduction
which gives the relevant syntax for these SME2 operations. 

My expectation (just a guess!) is that 
- use of ZA as target register rather than Z for SSVE ops will have a massive effect
- use of multi-vector loads and stores will have minimal performance effect (but will save some power; and will open up execute bandwidth for code that is doing other work on the CPU alongside the SME work)
- use of strided loads and ZA horizontal and vertical moves rather than element-by-element loads should have a noticeable effect

The other obvious way to improve column loads and transposes, should SME2 strided loads and/or use of the ZA register not be feasible, is by use of something like a NEON ld4 AoS<->SoA transform to rearrange the data via NEON before storing it to be loaded (as full 512b vectors) by SME. This can only be more performant than one element at a time loading! 

Another interesting idea you might want to investigate is the use of SME2 for flood fills. 
DC ZVA gives us a single cycle cache line (so 64B) zeroing, but the AMX/SME unit has a wider connection to L2, and may be able to zero (especially via VGx2 or VGx4) two 64B lines per cycle. And if you want to flood fill with a non-zero constant or simple repeating pattern, you may be able to get 3x or more the CPU flood fill performance?
Another case where this can be interesting is using AMX (as a collection of independently handled wide vectors) as a pseudo-random generator that can run almost independently and can probably keep up with many interesting use cases (unlike so many pseudo-random generator with act to throttle the code using them).

**Comment by breuera** [comment]
> Is the goal of these tests to show optimal performance, or performance in the hands of non-experts?
> Of course both types of tests have value.

It is not entirely clear to me what you mean by "tests". Based on the comments below, I assume that you are referring to our microbenchmarks. In that case: Most of them benchmark the throughput of a single instruction: We get the optimal performance for that particular instruction, possibly under additional constraints, such as RAW dependencies. I do not see an expert/non-expert separation at this point.

> use of Z rather than ZA as target register in the SSVE operations

I assume that this refers to the following code:
https://github.com/scalable-analyses/sme/blob/611a5a1e7c83ce7002e81e511d621395406e2479/kernels.s#L301-L337
Yes, we are fully aware that there are faster options for vector math in M4. For us this is future work as outlined here: https://scalable.uni-jena.de/opt/sme/micro.html#summary-and-outlook The reason for looking at FMLA (vectors) in the first place, is that all SVE GEMM implementations use it (or similar instructions).

> non-use of multi-vector (VG4x) loads and stores

Are you referring to these loads?

https://github.com/scalable-analyses/sme/blob/611a5a1e7c83ce7002e81e511d621395406e2479/kernels_gemm.s#L259-L263

> vastly sub-optimal construction of transposed matrices or column loads (should multi-vector strided load to Z, move to ZA horizontals, move back to Z as ZA verticals)

I am not aware that we have pushed (or written) any code for matrix transposition. We played around with zip4 in the vector registers, but that's about it.

> The other obvious way to improve column loads and transposes, should SME2 strided loads and/or use of the ZA register not be feasible, is by use of something like a NEON ld4 AoS<->SoA transform to rearrange the data via NEON before storing it to be loaded (as full 512b vectors) by SME. This can only be more performant than one element at a time loading!

The SME Programmer's Guide contains code for a non-transposed B. This will be our first direction when looking at transpositions in our GEMMS.

**Comment by name99-org** [comment]
> On May 26, 2024, at 11:56 PM, Alexander Breuer ***@***.***> wrote:
> 
> 
> Is the goal of these tests to show optimal performance, or performance in the hands of non-experts?
> Of course both types of tests have value.
> 
> It is not entirely clear to me what you mean by "tests". Based on the comments below, I assume that you are referring to our microbenchmarks. In that case: Most of them benchmark the throughput of a single instruction: We get the optimal performance for that particular instruction, possibly under additional constraints, such as RAW dependencies. I do not see an expert/non-expert separation at this point.
> 
> use of Z rather than ZA as target register in the SSVE operations
> 
> I assume that this refers to the following code:
> https://github.com/scalable-analyses/sme/blob/611a5a1e7c83ce7002e81e511d621395406e2479/kernels.s#L301-L337
> Yes, we are fully aware that there are faster options for vector math in M4. For us this is future work as outlined here: https://scalable.uni-jena.de/opt/sme/micro.html#summary-and-outlook The reason for looking at FMLA (vectors) in the first place, is that all SVE GEMM implementations use it (or similar instructions).
> 

The pattern I see is that your team is well aware of SVE/2, and some of SME, but it mostly unaware that Apple implements SME2, and that has some important differences. 
The best document available right now appears to be https://developer.arm.com/documentation/109246/0100/?lang=en although it is hardly complete. 

Regarding Z vs ZA, with SME a particular VECTOR instruction (not matrix instruction) like 
    fmla z0.s, p0/m, z30.s, z31.s
can take a ZA register (not a Z register) as the destination register. This is massively faster. If you understand how the AMX/SME hardware is laid out, this makes sense. 

Syntax something like 
	fmla za0.s[w8,0], p0/m, z30.s, z31.s 
where w8 is an index into the ZA register set, probably best set to 0 initially. (The entire ZA register addressing is something of a nightmare given how few bits are available).
> non-use of multi-vector (VG4x) loads and stores
> 
> Are you referring to these loads?
> 
> https://github.com/scalable-analyses/sme/blob/611a5a1e7c83ce7002e81e511d621395406e2479/kernels_gemm.s#L259-L263
> 


Likewise SME vector operations can take up to four vectors as each of the arguments. The syntax is now something like
    fmla za.s[w8, 0, VGx4], {z0.s-z3.s}, {z4.s-z7.s}
The use of the ZA destination makes a huge difference. The use of the {z0.s-z3.s} vector group to perform four multiply-adds has little SME performance advantage but uses much less CPU instruction bandwidth (which could then be used, for example, on helper NEON code). 

> vastly sub-optimal construction of transposed matrices or column loads (should multi-vector strided load to Z, move to ZA horizontals, move back to Z as ZA verticals)
> 
> I am not aware that we have pushed (or written) any code for matrix transposition. We played around with zip4 in the vector registers, but that's about it.
> 
The issue is not exactly matrix transposition, although that's sometimes required, it is the loading of the COLUMNS of the first of two matrices being multiplied. At some point you will need to do this, rather than simply loading a pre-transposed matrix. And the optimal way to do this is not obvious, especially if you are not using SME2 operations. 
This is discussed on page 59 of the PDF version of the link I sent you. They are discussing a fully general function so there's a lot more overhead and complexity, but the part I wanted to draw your attention to was 
- use VGx2 or VGx4 loads
- use the ZAH and ZAV register notation to load those stored values as rows, then read them as columns.

I really urge you to read the SME2 intro ASAP. The simple version is here: https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/arm-scalable-matrix-extension-introduction

You have performed a great service for the community by clarifying that SME is present on M4; but to really make this hardware shine, you need to use the full instruction set, and right now you are only using half of it! Which is fair, of course, no-one knew of this other set until a few days ago, except if they followed details on the LLVM lists or were deep into the instruction details of how AMX works. 

If you want to provide generic GEMMs and suchlike, of course you will need all the performance hints I've already described, but the one additional element that you will also need is the new counting-predicate mechanism, also described in https://developer.arm.com/documentation/109246/0100/?lang=en ; but that can wait until you have the simple cases like 16x16 and 32x32 optimized. 

Have fun!

**Comment by breuera** [comment]
> Regarding Z vs ZA, with SME a particular VECTOR instruction (not matrix instruction) like
    fmla z0.s, p0/m, z30.s, z31.s
can take a ZA register (not a Z register) as the destination register. This is massively faster. If you understand how the AMX/SME hardware is laid out, this makes sense.

As said before: We are fully aware that one can accumulate in ZA.
And once again: Your standard SVE GEMM library must be adopted to make use of this.

> The issue is not exactly matrix transposition, although that's sometimes required, it is the loading of the COLUMNS of the first of two matrices being multiplied.

Actually, we need to load (partial) rows from B, which is the harder case for column-major storage.

> At some point you will need to do this, rather than simply loading a pre-transposed matrix. And the optimal way to do this is not obvious, especially if you are not using SME2 operations.

While we will implement kernels for column-major A, B, and C, this depends on the use case. Especially, in ML inference, we can transpose the weight matrices up front.

> This is discussed on page 59 of the PDF version of the link I sent you. They are discussing a fully general function so there's a lot more overhead and complexity

Thank you for pointing this out, but as stated above - "The SME Programmer's Guide contains code for a non-transposed B." - we are already aware of this.

> Have fun!

To make this more productive: How about joining the effort?
Feel free to share your optimized GEMM routines.
We are interested in making our software fast, not in abstract discussions about which instruction should be microbenchmarked.

**Comment by name99-org** [comment]
I have done my part. I spent many many months analyzing how M1 AMX hardware works, and published this at 
https://github.com/name99-org/AArch64-Explore￼
name99-org/AArch64-Explore
github.com

(AMX is in volume 3. Updates will soon be providing giving 
- all changes across CPU, SoC and AMX since M1
- GPU
- ANE)
Corsix and Dougall used what I put together to do the next step, of disassembly.

I am exhausted. I just want people to understand when they are missing something, when they are using hardware sub-optimally because they don't understand the full story. 
But there are limits to what I can do in this respect; it's been astonishing (and immensely depressing) just how uninterested most people, even most supposed technophiles and engineers, are in fully understanding modern hardware beyond what's required to carry out tribal warfare.

**Comment by breuera** [comment]
All right, fair enough. Hope you find some new energy soon.

We still have a few weeks to go before we reach our goal of an SME JITter. At some point you'll probably even see some vector math for simpler primitives using the ZA tiles, it's just not a priority at the moment 😉
I am closing this issue. Feel free to reopen it if you see anything of interest in the future.

**Comment by name99-org** [comment]
Sorry I came across as curt. I am growing old and impatient – I want to see as much as possible of the story unfold before I die.
Regardless, I see that in your code you have mostly adopted the best practices I suggested, so great, progress! 
You're still not using the optimal vector instructions (as opposed to matrix instructions) but I assume that for now that's a lower priority than getting GEMM and friends done; when you do move to vectors remember what I said about using SSVE variants that write to ZA...

Point is, the code is better now, and that's what matters :-)
